From 5c58041ef4e1e8fddf948b0a9e768f3cacd95042 Mon Sep 17 00:00:00 2001
From: Romain Naour <romain.naour@gmail.com>
Date: Wed, 6 Aug 2025 22:32:30 +0200
Subject: [PATCH] microblaze fix: ira

Add new hooks for callee-save on microblaze defined by [1] and
return 1 to restore the old behavior prior to the commit [3].

Tested with qemu_microblazeel_mmu_defconfig

[1] https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=b191e8bdecf881d11c1544c441e38f4c18392a15
[2] https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/i386/i386.cc;h=3128973ba79cccfc6761f451dcb716b9558cc4da;hb=d3ff498c478acefce35de04402f99171b4f64a1a#l20606
[3] https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=3b9b8d6cfdf59337f4b7ce10ce92a98044b2657b

Upstream: under discussion at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=121432
Signed-off-by: Romain Naour <romain.naour@gmail.com>
---
 gcc/config/microblaze/microblaze.cc | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/gcc/config/microblaze/microblaze.cc b/gcc/config/microblaze/microblaze.cc
index f3ac94f0150..8ac52432bc9 100644
--- a/gcc/config/microblaze/microblaze.cc
+++ b/gcc/config/microblaze/microblaze.cc
@@ -3224,6 +3224,14 @@ microblaze_secondary_reload (bool in_p ATTRIBUTE_UNUSED, rtx x ATTRIBUTE_UNUSED,
   return NO_REGS;
 }
 
+/* Implement TARGET_CALLEE_SAVE_COST.  */
+static int
+microblaze_callee_save_cost (spill_cost_type, unsigned int hard_regno, machine_mode,
+		       unsigned int, int mem_cost, const HARD_REG_SET &, bool)
+{
+  return 1;
+}
+
 static void
 microblaze_globalize_label (FILE * stream, const char *name)
 {
@@ -4062,6 +4070,9 @@ microblaze_starting_frame_offset (void)
 #undef TARGET_SECONDARY_RELOAD
 #define TARGET_SECONDARY_RELOAD		microblaze_secondary_reload
 
+#undef TARGET_CALLEE_SAVE_COST
+#define TARGET_CALLEE_SAVE_COST microblaze_callee_save_cost
+
 #undef  TARGET_ASM_OUTPUT_MI_THUNK
 #define TARGET_ASM_OUTPUT_MI_THUNK      microblaze_asm_output_mi_thunk
 
-- 
2.50.1

